src/lib.rs

```diff
@@ -77,24 +77,41 @@ impl<IT, N> Key<IT, N> {
     }
 }
 
+/// Module to hide the Entry type which needs to be public due to the generic-array internals.
+mod entry {
+    pub enum Entry<IT> {
+        Used(IT),
+        EmptyNext(usize),
+        EmptyLast
+    }
+}
+
+use entry::Entry;
+
 // Data type that stores values and returns a key that can be used to manipulate
 // the stored values.
 // Values can be read by anyone but can only be modified using the key.
 pub struct Slots<IT, N>
-    where N: ArrayLength<Option<IT>> + ArrayLength<usize> + Unsigned {
-    items: GenericArray<Option<IT>, N>,
-    free_list: GenericArray<usize, N>,
+    where N: ArrayLength<Entry<IT>> + Unsigned {
+    items: GenericArray<Entry<IT>, N>,
+    // Could be optimized by making it just usize and relying on free_count to determine its
+    // validity
+    next_free: Option<usize>,
     free_count: usize
 }
 
 impl<IT, N> Slots<IT, N>
-    where N: ArrayLength<Option<IT>> + ArrayLength<usize> + Unsigned {
+    where N: ArrayLength<Entry<IT>> + Unsigned {
     pub fn new() -> Self {
         let size = N::to_usize();
 
         Self {
-            items: GenericArray::default(),
-            free_list: GenericArray::generate(|i: usize| size - i - 1),
+            // Fill back-to-front
+            // items: GenericArray::generate(|i: usize| if i == 0 { Entry::EmptyLast } else { Entry::EmptyNext(i - 1) }),
+            // next_free: size.checked_sub(1),
+            // Fill front-to-back
+            items: GenericArray::generate(|i: usize| if i == size - 1 { Entry::EmptyLast } else { Entry::EmptyNext(i + 1) }),
+            next_free: Some(0),
```

Comment:
I don't mind either solution but please don't leave commented code in the codebase.

Reply:
Picked and simplified the former; that changes the indices generated for the first items, but I don't see any guarantees on them being dealt out in a particular order.